Add support for ISO-2022-JP encoding in multipart form data handling(#2245)

nappa opened the pull request:
Hi, I introduced simple workaround for ISO-2022-JP encoding in multipart form data handling, which solves #2245 .
Also, I added test case to verify correct handling of multipart data with ISO-2022-JP encoding.

I hope it helps.

ioquatix commented on lib/rack/multipart/parser.rb:
```diff
@@ -486,6 +487,19 @@ def find_encoding(enc)
+        if name.encoding.dummy?
+          # ISO-2022-JP is a legacy but still widely used encoding in Japan
+          # Here we properly convert ISO-2022-JP to UTF-8
+          if name.encoding == Encoding::ISO_2022_JP
+            name = name.encode(Encoding::UTF_8)
+            body = body.encode(Encoding::UTF_8)
+          end
```
Why not handle all dummy encodings thsi way, e.g.

```suggestion
          name = name.encode(Encoding::UTF_8)
          body = body.encode(Encoding::UTF_8)
```

ioquatix commented:
For reference:

```
irb(main):002> Encoding.list.select{|encoding| encoding.dummy?}
=> 
[#<Encoding:UTF-16 (dummy)>,
 #<Encoding:UTF-32 (dummy)>,
 #<Encoding:IBM037 (dummy)>,
 #<Encoding:ISO-2022-JP (dummy)>,
 #<Encoding:ISO-2022-JP-2 (dummy)>,
 #<Encoding:CP50220 (dummy)>,
 #<Encoding:CP50221 (dummy)>,
 #<Encoding:UTF-7 (dummy)>,
 #<Encoding:ISO-2022-JP-KDDI (dummy)>]
```

jeremyevans commented:
I don't know much about dummy encodings, but if they aren't really usable, encoding to UTF-8 seems reasonable.  It seems more consistent to do this for all dummy encodings, so @ioquatix's suggestion makes sense.  CHANGELOG should be updated when using @ioquatix's suggestion.

Security-wise, this would expose us to any issues in Ruby implementation transcoders from dummy encodings to UTF-8, as it would allow transcoding attacker-supplied data from an attacker-supplied dummy encoding to UTF-8.  Not knowing much about dummy encodings, I cannot estimate the actual risk.

ioquatix commented:
@duerst are you able to help us review this PR, and can you consider Jeremy's concerns/feedback?

nappa commented:
Thank you so much for taking the time to view this!

The reason I chose to exclude dummy encodings other than ISO-2022-JP in the `if` statement is that I considered whether they are actually used on the internet. Let me explain the situation with dummy encodings other than ISO-2022-JP:

Among the dummy encodings, CP50220, CP50221, ISO-2022-JP-KDDI, and ISO-2022-JP-2 are all variants of ISO-2022-JP.

- **CP50220**: This is an extension of ISO-2022-JP by Microsoft. It is registered in the IANA Registry. The registration was requested by one of the Ruby developers. [Source](https://data.iana.org/archive/ietf-charsets/msg01882.html) It is used only by a very limited number of applications and is now rarely seen on the internet.  
- **CP50221**: This is also an extension of ISO-2022-JP by Microsoft, but it is *not* registered in the IANA Registry. It is now rarely seen on the internet.  
- **ISO-2022-JP-2**: This is defined in RFC 1554 and is registered in the IANA Registry. However, it is now rarely seen on the internet.  
- **ISO-2022-JP-KDDI**: This is used by older devices from the Japanese mobile carrier KDDI. It is *not* registered in the IANA Registry. Since ISO-2022-JP is not specified in the `Content-Type`, it is necessary to infer that the device is a KDDI mobile phone based on information like the `User-Agent` and then explicitly specify ISO-2022-JP-KDDI for conversion.

Other encodings are as follows:  

- **UTF-16, UTF-32**: To convert these to UTF-8, determining the byte order is necessary. If a BOM (Byte Order Mark) is included at the beginning of the text, it can be processed correctly. However, if a BOM is not included, an `Encoding::InvalidByteSequenceError` will be raised. Attempting conversion in such cases may lead to other issues.  
- **IBM037**: I have never seen this encoding used on the internet, and I believe it is highly unlikely to appear.

Frankly speaking, the considerations for encodings other than ISO-2022-JP are far too numerous, and they also carry a significant potential for causing additional issues. Therefore, narrowing the scope to only ISO-2022-JP is not a bad choice from a practical standpoint.

duerst commented:
I agree with the direction this is going. At the moment, indeed Encoding::ISO_2022_JP seems to be the only encoding where this special treatment is needed and appropriate.  As for security issues, I think the main concern would be about smuggling some special syntax in the iso-2022-jp that then triggers something like Bobby Tables (https://xkcd.com/327/). That shouldn't happen because in iso-2022-jp, all ASCII characters appear literally. It also shouldn't happen because the security check (or other security measure, such as additional quoting/escaping) should be applied after the encoding step, not before. The multipart/form-data may contain other kinds of encoding on top of iso-2022-jp (e.g. quoted-printable or base64) that hide things further, but that's also possible for all the non-dummy encodings.

jeremyevans reviewed (approved):
Based on @nappa's and @duerst's feedback, I think we should accept this as-is.  I recommend we update the CHANGELOG to specify that ISO-2022-JP is transcoded to UTF-8, and I'd be OK with switching to a hash-table approach for more user control, as recommended by @ioquatix.

nappa commented:
Thank you so much, everyone!
Based on @ioquatix's feedback, I added a Hash table and included comments. I also updated the CHANGELOG.md following @jeremyevans's suggestions. Please let me know if there's anything else I should work on!

jeremyevans reviewed (approved):
Looks good.  Thank you for the quick fix!

ioquatix commented:
I’ll review this tonight in a few hours.

ioquatix commented:
Thanks everyone.

nappa commented:
@ioquatix @jeremyevans @duerst I'm happy to have been able to support this project. Thank you so much!
